src/python/pants/backend/experimental/python/lockfile.py

```diff
@@ -51,6 +58,75 @@ def lockfile_dest(self) -> str:
         return cast(str, self.options.lockfile_dest)
 
 
+@dataclass(frozen=True)
+class PythonLockfile:
+    digest: Digest
+    path: str
+
+
+@dataclass(frozen=True)
+class PythonLockfileRequest:
+    requirements: FrozenOrderedSet[str, ...]
+    interpreter_constraints: InterpreterConstraints
+    dest: str
+    description: str
+
+    @classmethod
+    def from_tool(cls, subsystem: PythonToolRequirementsBase) -> PythonLockfileRequest:
+        return cls(
+            requirements=FrozenOrderedSet(subsystem.all_requirements),
+            interpreter_constraints=subsystem.interpreter_constraints,
+            dest=subsystem.lockfile,
+            description=f"Generate lockfile for {subsystem.options_scope}",
+        )
+
+
+@rule(desc="Generate lockfile", level=LogLevel.DEBUG)
+async def generate_lockfile(
+    req: PythonLockfileRequest, pip_tools_subsystem: PipToolsSubsystem
+) -> PythonLockfile:
+    input_requirements = await Get(
+        Digest, CreateDigest([FileContent("reqs.txt", "\n".join(req.requirements).encode())])
```

Comment:
Note that this file name will be included in the generated lockfile like this:

https://github.com/pantsbuild/pants/blob/b5781b5601afe372dcd14d6cc041a912e32d4d39/src/python/pants/backend/python/lint/black/lockfile.txt#L11-L14

This is somewhat of a UX regression to always use `reqs.txt`, but I figure it might in some ways be less confusing than the old naming schemes we were using? Thoughts on an optimal UX? I'd rather prioritize good UX than simpler code here.

Note that we'll be changing the lockfile header already to share information about how the lockfile can be regenerated, along w/ maybe things like which ICs were used.

https://github.com/pantsbuild/pants/blob/b5781b5601afe372dcd14d6cc041a912e32d4d39/src/python/pants/backend/python/lint/black/lockfile.txt#L1-L6